barcodes_search/wizards/barcode_search.py

```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2017-Today GRAP (http://www.grap.coop)
+# @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class BarcodeSearch(models.TransientModel):
+    _name = 'barcode.search'
+
+    barcode = fields.Char(
+        string='Barcode', required=True, help='Barcode to search for.')
+
+    state = fields.Selection(
+        selection=[('new', 'New'), ('done', 'Done')], default='new')
+
+    line_ids = fields.One2many(
+        comodel_name='barcode.search.line', inverse_name='barcode_search_id',
+        readonly=True)
+
+    @api.multi
+    def button_search(self):
+        self.ensure_one()
+        res = self.search_by_barcode(self.barcode)
+        if len(res) == 1 and not res[0][2]:
+            # Display the result in the according form view
+            record = res[0][1]
+            return record.get_formview_action()
+        else:
+            # Update Wizard Values
+            line_vals = []
+            self.line_ids.unlink()
+            for item in res:
+                line_vals.append((0, 0, {
+                    'field_id': item[0].id,
+                    'item_id': item[1].id,
+                    'item_name': item[1].name,
+                    'extra_data': item[2],
+                }))
+            self.state = 'done'
+            self.line_ids = line_vals
+
+            # Return Action
+            action = self.get_formview_action()
+            action['target'] = 'new'
+            return action
+
+    @api.model
+    def search_by_barcode(self, barcode):
+        """Return a list of tuple (field, record, {extra_data})
```

Comment:
Hmmmm the tuple structure is really hard to follow IMO. Any reason you didn't use a frozendict or namedtuple?

As evidence of this, I had to jump down here while reading the L26 if statement in order to ascertain what that statement does.